cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java

```diff
@@ -113,18 +118,95 @@ public ExecutionSparkContext(SparkSpecification specification, Id.Program progra
 
   @Override
   public <T> T readFromDataset(String datasetName, Class<?> kClass, Class<?> vClass) {
-    Map<String, String> datasetArgs = RuntimeArguments.extractScope(Scope.DATASET, datasetName, getRuntimeArguments());
-
     // Clone the configuration since it's dataset specification and shouldn't affect the global hConf
     Configuration configuration = new Configuration(hConf);
+
+    // first try if it is InputFormatProvider
+    Map<String, String> dsArgs = RuntimeArguments.extractScope(Scope.DATASET, datasetName, getRuntimeArguments());
+    Dataset dataset = getDataset(datasetName, dsArgs);
```

Comment:
So, the getDatasets() remembers all datasets it instantiates in the datasets Map. If this is iterative, and the same dataset is used again, getDataset() will return the same instance. Which would be bad if I close it here. So, if I close it, I have to remove it from the datasets Map, for which currently there is no way provided. But if I do that, then rollbackTx() won't be called in case the program fails later. So I believe I have to keep it open.